crates/vm/execution_db.rs

```diff
@@ -0,0 +1,146 @@
+};
+
+/// In-memory EVM database for caching execution data.
+///
+/// This is mainly used to store the relevant state data for executing a particular block and then
+/// feeding the DB into a zkVM program to prove the execution.
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct ExecutionDB {
+    /// indexed by account address
+    accounts: HashMap<RevmAddress, RevmAccountInfo>,
+    /// indexed by code hash
+    code: HashMap<RevmB256, RevmBytecode>,
+    /// indexed by account address and storage slot
+    storage: HashMap<RevmAddress, HashMap<RevmU256, RevmU256>>,
+    /// indexed by block number
+    block_hashes: HashMap<u64, RevmB256>,
+    /// stored chain config
+    chain_config: ChainConfig,
+}
+
+impl ExecutionDB {
+    /// Creates a database by executing a block, without performing any validation.
+    pub fn from_exec(block: &Block, store: &Store) -> Result<Self, ExecutionDBError> {
+        // TODO: perform validation to exit early
+
+        let mut state = evm_state(store.clone(), block.header.parent_hash);
+        let mut store_wrapper = StoreWrapper {
+            store: store.clone(),
+            block_hash: block.header.parent_hash,
+        };
+
+        let chain_config = store.get_chain_config()?;
+
+        execute_block(block, &mut state).map_err(Box::new)?;
+
+        let account_updates = get_state_transitions(&mut state);
+
+        let mut accounts = HashMap::new();
+        let code = HashMap::new(); // TODO: `code` remains empty for now
+        let mut storage = HashMap::new();
+        let mut block_hashes = HashMap::new();
+
+        for account_update in account_updates {
+            let address = RevmAddress::from_slice(account_update.address.as_bytes());
+            let account_info = store_wrapper
+                .basic(address)?
+                .ok_or(ExecutionDBError::NewMissingAccountInfo(address))?;
+            accounts.insert(address, account_info);
+
+            let account_storage = account_update
+                .added_storage
+                .iter()
+                .map(|(slot, value)| {
+                    let mut value_bytes = [0u8; 32];
+                    value.to_big_endian(&mut value_bytes);
+                    (
+                        RevmU256::from_be_bytes(slot.to_fixed_bytes()),
+                        RevmU256::from_be_slice(&value_bytes),
+                    )
+                })
+                .collect();
+
+            storage.insert(address, account_storage);
+        }
+
+        let earliest_number = store_wrapper
+            .store
+            .get_earliest_block_number()?
+            .ok_or(ExecutionDBError::NewMissingBlockNumber())?;
+        let latest_number = store_wrapper
+            .store
+            .get_latest_block_number()?
+            .ok_or(ExecutionDBError::NewMissingBlockNumber())?;
+
+        let numbers = earliest_number..=latest_number;
```

Comment:
Why is this needed? This doesn't scale well, since for every new block you will have to do a new db read.

Reply:
What you mean is that I'm reading the hashes of all blocks every time from the DB instead of retrieving them once and then adding each new hash to the list, right?

Reply:
In mainnet there are 21M blocks since genesis..

Reply:
Yes, it's not optimized, but it’s not intended for the mainnet. We require all prior blocks **strictly necessary** to execute any desired block. The goal is to create the smallest possible "database", that can be de/serialiazable, to facilitate the `execute_block` operation for the prover. Since we're currently testing with L2 and have only a few blocks, this approach is feasible.

Reply:
i've created an issue to track this problem/optimization.
#1016

Reply:
Ok

Reply:
just applied that change.